File: VB/WinWebSolution.Module/Activity.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Xml
Imports DevExpress.Xpo
Imports DevExpress.ExpressApp
Imports System.ComponentModel
Imports DevExpress.Xpo.Metadata
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.Persistent.Base.General

Namespace WinWebSolution.Module
	<DefaultClassOptions> _
	Public Class Activity
		Inherits BaseObject
		Implements IEvent, ISupportRecurrences
		Private _AllDay As Boolean
		Private _Description As String
		Private _StartOn As DateTime
		Private _EndOn As DateTime
		Private _Label As Integer
		Private _Location As String
		Private _Status As Integer
		Private _Subject As String
		Private _Type As Integer
		<Persistent("ResourceIds"), Size(SizeAttribute.Unlimited)> _
		Private _EmployeeIds As String
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub
		Public Property AllDay() As Boolean Implements IEvent.AllDay
			Get
				Return Me._AllDay
			End Get

			Set(ByVal value As Boolean)
				SetPropertyValue("AllDay", Me._AllDay, value)
			End Set
		End Property
		<NonPersistent, Browsable(False)> _
		Public ReadOnly Property AppointmentId() As String Implements IEvent.AppointmentId
			Get
				Return Me.Oid.ToString()
			End Get
		End Property
		<Size(SizeAttribute.Unlimited)> _
		Public Property Description() As String Implements IEvent.Description
			Get
				Return Me._Description
			End Get

			Set(ByVal value As String)
				SetPropertyValue("Description", Me._Description, value)
			End Set
		End Property
		Public Property Label() As Integer Implements IEvent.Label
			Get
				Return Me._Label
			End Get

			Set(ByVal value As Integer)
				SetPropertyValue("Label", Me._Label, value)
			End Set
		End Property
		Public Property Location() As String Implements IEvent.Location
			Get
				Return Me._Location
			End Get

			Set(ByVal value As String)
				SetPropertyValue("Location", Me._Location, value)
			End Set
		End Property
		<PersistentAlias("_EmployeeIds"), Browsable(False)> _
		Public Property ResourceId() As String Implements IEvent.ResourceId
			Get
				If Me._EmployeeIds Is Nothing Then
					Me.UpdateEmployeeIds()
				End If
				Return Me._EmployeeIds
			End Get
			Set(ByVal value As String)
				If _EmployeeIds <> value AndAlso value IsNot Nothing Then
					Me._EmployeeIds = value
					Me.UpdateEmployees()
				End If
			End Set
		End Property
		<Indexed> _
		Public Property StartOn() As DateTime Implements IEvent.StartOn
			Get
				Return Me._StartOn
			End Get

			Set(ByVal value As DateTime)
				SetPropertyValue("StartOn", Me._StartOn, value)
			End Set
		End Property
		<Indexed> _
		Public Property EndOn() As DateTime Implements IEvent.EndOn
			Get
				Return Me._EndOn
			End Get

			Set(ByVal value As DateTime)
				SetPropertyValue("EndOn", Me._EndOn, value)
			End Set
		End Property
		Public Property Status() As Integer Implements IEvent.Status
			Get
				Return Me._Status
			End Get

			Set(ByVal value As Integer)
				SetPropertyValue("Status", Me._Status, value)
			End Set
		End Property
		<Size(250)> _
		Public Property Subject() As String Implements IEvent.Subject
			Get
				Return Me._Subject
			End Get

			Set(ByVal value As String)
				SetPropertyValue("Subject", Me._Subject, value)
			End Set
		End Property
		<Browsable(False)> _
		Public Property Type() As Integer Implements IEvent.Type
			Get
				Return Me._Type
			End Get

			Set(ByVal value As Integer)
				SetPropertyValue("Type", Me._Type, value)
			End Set
		End Property
		<Association("Activity-Employees", UseAssociationNameAsIntermediateTableName := True)> _
		Public ReadOnly Property Employees() As XPCollection(Of Employee)
			Get
				Return GetCollection(Of Employee)("Employees")
			End Get
		End Property
		Public Overrides Sub AfterConstruction()
			MyBase.AfterConstruction()
			Me.StartOn = DateTime.Now
			Me.EndOn = Me.StartOn.AddHours(1)
			Me.Employees.Add(Session.GetObjectByKey(Of Employee)(SecuritySystem.CurrentUserId))
		End Sub
		Public Sub UpdateEmployeeIds()
			Me._EmployeeIds = String.Empty
			Me.Employees.SuspendChangedEvents()
			Try
				For Each activityUser As Employee In Me.Employees
					Me._EmployeeIds &= String.Format("<ResourceId Type=""{0}"" Value=""{1}"" />", activityUser.Id.GetType().FullName, activityUser.Id)
				Next activityUser
			Finally
				Me.Employees.ResumeChangedEvents()
			End Try
			Me._EmployeeIds = "<ResourceIds>" & Me._EmployeeIds & "</ResourceIds>"
		End Sub
		Protected Overrides Overloads Function CreateCollection(Of T)(ByVal [property] As XPMemberInfo) As XPCollection(Of T)
			Dim result As XPCollection(Of T) = MyBase.CreateCollection(Of T)([property])
			If [property].Name = "Employees" Then
				AddHandler result.ListChanged, AddressOf Employees_ListChanged
			End If
			Return result
		End Function
		Private Sub UpdateEmployees()
			Me.Employees.SuspendChangedEvents()
			Try
				Do While Me.Employees.Count > 0
					Me.Employees.Remove(Me.Employees(0))
				Loop
				If (Not String.IsNullOrEmpty(_EmployeeIds)) Then
					Dim xmlDocument As New XmlDocument()
					xmlDocument.LoadXml(_EmployeeIds)
					For Each xmlNode As XmlNode In xmlDocument.DocumentElement.ChildNodes
						Dim activityUser As Employee = Session.GetObjectByKey(Of Employee)(New Guid(xmlNode.Attributes("Value").Value))
						If activityUser IsNot Nothing Then
							Me.Employees.Add(activityUser)
						End If
					Next xmlNode
				End If
			Finally
				Me.Employees.ResumeChangedEvents()
			End Try
		End Sub
		Private Sub Employees_ListChanged(ByVal sender As Object, ByVal e As ListChangedEventArgs)
			If e.ListChangedType = ListChangedType.ItemAdded OrElse e.ListChangedType = ListChangedType.ItemDeleted Then
				UpdateEmployeeIds()
				OnChanged("ResourceId")
			End If
		End Sub
		Private recurrenceInfoXml_Renamed As String
		<DevExpress.Xpo.DisplayName("Recurrence"), Size(SizeAttribute.Unlimited)> _
		Public Property RecurrenceInfoXml() As String Implements ISupportRecurrences.RecurrenceInfoXml
			Get
				Return recurrenceInfoXml_Renamed
			End Get
			Set(ByVal value As String)
				If recurrenceInfoXml_Renamed <> value Then
					recurrenceInfoXml_Renamed = value
					OnChanged("RecurrenceInfoXml")
				End If
			End Set
		End Property
		<Persistent("RecurrencePattern")> _
		Private recurrencePattern_Renamed As Activity
		Public Property RecurrencePattern() As IEvent Implements ISupportRecurrences.RecurrencePattern
			Get
				Return recurrencePattern_Renamed
			End Get
			Set(ByVal value As IEvent)
				recurrencePattern_Renamed = CType(value, Activity)
				OnChanged("RecurrencePattern")
			End Set
		End Property
		Protected Overrides Sub OnLoaded()
			MyBase.OnLoaded()
			If Employees.IsLoaded Then
				Employees.Reload()
			End If
		End Sub
	End Class
End Namespace

File: CS/WinWebSolution.Module/Activity.cs
using System;
using System.Xml;
using DevExpress.Persistent.Base;
using DevExpress.Persistent.Base.General;
using DevExpress.Persistent.BaseImpl;
using DevExpress.Xpo;
using DevExpress.Xpo.Metadata;
using DevExpress.ExpressApp;
using System.ComponentModel;

namespace WinWebSolution.Module {
    [DefaultClassOptions]
    public class Activity : BaseObject, IEvent, ISupportRecurrences {
        private bool _AllDay;
        private string _Description;
        private DateTime _StartOn;
        private DateTime _EndOn;
        private int _Label;
        private string _Location;
        private int _Status;
        private string _Subject;
        private int _Type;
        [Persistent("ResourceIds"), Size(SizeAttribute.Unlimited), ObjectValidatorIgnoreIssue(typeof(ObjectValidatorLargeNonDelayedMember))]
        private string _EmployeeIds;
        public Activity(Session session)
            : base(session) {
        }
        public bool AllDay {
            get {
                return this._AllDay;
            }

            set {
                SetPropertyValue("AllDay", ref this._AllDay, value);
            }
        }
        [NonPersistent]
        public string AppointmentId {
            get { return this.Oid.ToString(); }
        }
        [Size(SizeAttribute.Unlimited), ObjectValidatorIgnoreIssue(typeof(ObjectValidatorLargeNonDelayedMember))]
        public string Description {
            get {
                return this._Description;
            }

            set {
                SetPropertyValue("Description", ref this._Description, value);
            }
        }
        public int Label {
            get {
                return this._Label;
            }

            set {
                SetPropertyValue("Label", ref this._Label, value);
            }
        }
        public string Location {
            get {
                return this._Location;
            }

            set {
                SetPropertyValue("Location", ref this._Location, value);
            }
        }
        [Browsable(false)]
        [PersistentAlias("_EmployeeIds")]
        public string ResourceId {
            get {
                if (this._EmployeeIds == null) {
                    this.UpdateEmployeeIds();
                }
                return this._EmployeeIds;
            }
            set {
                if (_EmployeeIds != value && value != null) {
                    this._EmployeeIds = value;
                    this.UpdateEmployees();
                }
            }
        }
        [Indexed]
        public DateTime StartOn {
            get {
                return this._StartOn;
            }

            set {
                SetPropertyValue("StartOn", ref this._StartOn, value);
            }
        }
        [Indexed]
        public DateTime EndOn {
            get {
                return this._EndOn;
            }

            set {
                SetPropertyValue("EndOn", ref this._EndOn, value);
            }
        }
        public int Status {
            get {
                return this._Status;
            }

            set {
                SetPropertyValue("Status", ref this._Status, value);
            }
        }
        [Size(250)]
        public string Subject {
            get {
                return this._Subject;
            }

            set {
                SetPropertyValue("Subject", ref this._Subject, value);
            }
        }
        [Browsable(false)]
        public int Type {
            get {
                return this._Type;
            }

            set {
                SetPropertyValue("Type", ref this._Type, value);
            }
        }
        [Association("Activity-Employees", typeof(Employee), UseAssociationNameAsIntermediateTableName = true)]
        public XPCollection Employees {
            get { return GetCollection("Employees"); }
        }
        public override void AfterConstruction() {
            base.AfterConstruction();
            this.StartOn = DateTime.Now;
            this.EndOn = this.StartOn.AddHours(1);
            this.Employees.Add(Session.GetObjectByKey<Employee>(SecuritySystem.CurrentUserId));
        }
        public void UpdateEmployeeIds() {
            this._EmployeeIds = string.Empty;
            this.Employees.SuspendChangedEvents();
            try {
                foreach (Employee activityUser in this.Employees) {
                    this._EmployeeIds += String.Format(@"<ResourceId Type=""{0}"" Value=""{1}"" />", activityUser.Id.GetType().FullName, activityUser.Id);
                }
            } finally {
                this.Employees.ResumeChangedEvents();
            }
            this._EmployeeIds = "<ResourceIds>" + this._EmployeeIds + "</ResourceIds>";
        }
        protected override XPCollection CreateCollection(XPMemberInfo property) {
            XPCollection result = base.CreateCollection(property);
            if (property.Name == "Employees") {
                result.CollectionChanged += new XPCollectionChangedEventHandler(this.Employees_CollectionChanged);
            }

            return result;
        }
        private void UpdateEmployees() {
            this.Employees.SuspendChangedEvents();
            try {
                while (this.Employees.Count > 0) {
                    this.Employees.Remove(this.Employees[0]);
                }
                if (!String.IsNullOrEmpty(this._EmployeeIds)) {
                    XmlDocument xmlDocument = new XmlDocument();
                    xmlDocument.LoadXml(this._EmployeeIds);
                    foreach (XmlNode xmlNode in xmlDocument.DocumentElement.ChildNodes) {
                        Employee activityUser = Session.GetObjectByKey<Employee>(new Guid(xmlNode.Attributes["Value"].Value));
                        if (activityUser != null) {
                            this.Employees.Add(activityUser);
                        }
                    }
                }
            } finally {
                this.Employees.ResumeChangedEvents();
            }
        }
        private void Employees_CollectionChanged(object sender, XPCollectionChangedEventArgs e) {
            if ((e.CollectionChangedType == XPCollectionChangedType.AfterAdd) || (e.CollectionChangedType == XPCollectionChangedType.AfterRemove)) {
                this.UpdateEmployeeIds();
                OnChanged("ResourceId");
            }
        }
        private string recurrenceInfoXml;
        [DevExpress.Xpo.DisplayName("Recurrence"), Size(SizeAttribute.Unlimited), ObjectValidatorIgnoreIssue(typeof(ObjectValidatorLargeNonDelayedMember))]
        public string RecurrenceInfoXml {
            get {
                return recurrenceInfoXml;
            }
            set {
                recurrenceInfoXml = value;
                OnChanged("RecurrenceInfoXml");
            }
        }
        [Persistent("RecurrencePattern")]
        private Activity recurrencePattern;
        public IEvent RecurrencePattern {
            get {
                return recurrencePattern;
            }
            set {
                recurrencePattern = (Activity)value;
                OnChanged("RecurrencePattern");
            }
        }
    }
}